bundler/lib/bundler/checksum.rb

```diff
@@ -2,258 +2,174 @@
+      def register_gem_package(spec, source)
+        new_checksums = Checksum.from_gem_source(source)
+        new_checksums.each do |checksum|
+          register spec.full_name, checksum
         end
+      rescue SecurityError
+        expected = checksums(spec.full_name)
+        gem_lock_name = GemHelpers.lock_name(spec.name, spec.version, spec.platform)
+        raise SecurityError, <<~MESSAGE
+          Bundler cannot continue installing #{gem_lock_name}.
+          The checksum for the downloaded `#{spec.full_name}.gem` does not match \
+          the known checksum for the gem.
+          This means the contents of the downloaded \
+          gem is different from what was uploaded to the server \
+          or first used by your teammates, and could be a potential security issue.
 
-        io.rewind
-      end
+          To resolve this issue:
+          1. delete the downloaded gem located at: `#{source.path}`
+          2. run `bundle install`
 
-      digests
-    end
+          If you are sure that the new checksum is correct, you can \
+          remove the `#{gem_lock_name}` entry under the lockfile `CHECKSUMS` \
+          section and rerun `bundle install`.
 
-    def full_name
-      GemHelpers.spec_full_name(@name, @version, @platform)
-    end
+          If you wish to continue installing the downloaded gem, and are certain it does not pose a \
+          security issue despite the mismatching checksum, do the following:
+          1. run `bundle config set --local disable_checksum_validation true` to turn off checksum verification
+          2. run `bundle install`
 
-    def match_spec?(spec)
-      name == spec.name &&
-        version == spec.version &&
-        platform.to_s == spec.platform.to_s
-    end
+          #{expected.map do |checksum|
+            next unless actual = new_checksums.find {|c| c.algo == checksum.algo }
+            next if actual.digest == checksum.digest
 
-    def to_lock
-      out = String.new
-      out << "  #{GemHelpers.lock_name(name, version, platform)}"
-      checksums.sort_by(&:first).each_with_index do |(algo, checksum), idx|
-        out << (idx.zero? ? " " : ",")
-        out << algo << "-" << checksum
+            "(More info: The expected #{checksum.algo.upcase} checksum was #{checksum.digest.inspect}, but the " \
+            "checksum for the downloaded gem was #{actual.digest.inspect}. The expected checksum came from: #{checksum.sources.join(", ")})"
+          end.compact.join("\n")}
+        MESSAGE
       end
-      out << "\n"
 
-      out
-    end
+      def register(full_name, checksum)
+        return unless checksum
 
-    def match?(other)
-      return false unless match_spec?(other)
-      match_digests?(other.checksums)
-    end
+        sums = (store[full_name] ||= [])
+        sums.find {|c| c.algo == checksum.algo }&.merge!(checksum) || sums << checksum
+      rescue SecurityError => e
+        raise e.exception(<<~MESSAGE)
```

Comment:
can we extract out this message construction to a method, since we do it 2x in this file?

Reply:
I was working towards this. They are quite different. In the RFC I think I covered the cases where it will happen and wrote a simpler version. We could leave this as a task to do because it's linked to the entire way we verify checksums and how we message mismatches. (in other words, it's not staying like this, but the diff is so big it's hard to read and I'd like to make focused PRs rather than add commits on this one)